lib/charms/kubernetes_charm_libraries/v0/kubernetes_hugepages_volumes_patch.py

```diff
@@ -0,0 +1,624 @@
+
+
+class K8sHugePagesVolumePatchChangedCharmEvents(CharmEvents):
+    hugepages_volumes_config_changed = EventSource(K8sHugePagesVolumePatchChangedEvent)
+
+
+class YourCharm(CharmBase):
+
+    on = K8sHugePagesVolumePatchChangedCharmEvents()
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        self._kubernetes_volumes_patch = KubernetesHugePagesPatchCharmLib(
+            charm=self,
+            container_name=self._container_name,
+            hugepages_volumes_func=self._hugepages_volumes_func_from_config,
+            refresh_event=self.on.hugepages_volumes_config_changed,
+        )
+
+    def _hugepages_volumes_func_from_config(self) -> list[RequestedHugePages]:
+        return [
+            RequestedHugePages(
+                mount_path="/dev/hugepages",
+                size="1Gi",
+                limit="4Gi",
+            )
+        ]
+
+"""
+import logging
+from dataclasses import dataclass
+from typing import Callable, Iterable
+
+from lightkube import Client
+from lightkube.core.exceptions import ApiError
+from lightkube.models.apps_v1 import StatefulSetSpec
+from lightkube.models.core_v1 import (
+    Container,
+    EmptyDirVolumeSource,
+    ResourceRequirements,
+    Volume,
+    VolumeMount,
+)
+from lightkube.resources.apps_v1 import StatefulSet
+from lightkube.resources.core_v1 import Pod
+from ops.charm import CharmBase
+from ops.framework import BoundEvent, Object
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class RequestedHugePages:
+    """RequestedHugePages."""
+
+    mount_path: str
+    size: str = "1Gi"
+    limit: str = "2Gi"
+
+
+class KubernetesHugePagesVolumesPatchError(Exception):
+    """KubernetesHugePagesVolumesPatchError."""
+
+    def __init__(self, message: str):
+        self.message = message
+        super().__init__(self.message)
+
+
+class ContainerNotFoundError(ValueError):
+    """Raised when a given container does not exist in the iterable of containers."""
+
+
+class KubernetesClient:
+    """Class containing all the Kubernetes specific calls."""
+
+    def __init__(self, namespace: str):
+        self.client = Client()
+        self.namespace = namespace
+
+    @classmethod
+    def _get_container(cls, container_name: str, containers: Iterable[Container]) -> Container:
+        """Find the container from the container list, assuming list is unique by name.
+
+        Args:
+            containers: Iterable of containers
+            container_name: Container name
+
+        Raises:
+            ContainerNotFoundError, if the user-provided container name does not exist in the list.
+
+        Returns:
+            Container: An instance of :class:`Container` whose name matches the given name.
+        """
+        try:
+            return next(iter(filter(lambda ctr: ctr.name == container_name, containers)))
+        except StopIteration:
+            raise ContainerNotFoundError(f"Container `{container_name}` not found")
```

Comment:
Do we need a separate error for the container? All other lookup failures raise a generic `KubernetesHugePagesVolumesPatchError`?